superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+
+## Embedded RPC-based implementation
+
+As alternative to running a full `op-resolver` service, we may support a fallback in the op-node.
+Message safety is checked by relying on the local execution-engine
+for serving full event-logs and block-safety labels of the local chain,
+and then attaching to another external backend 
+(possibly served by an op-node / execution-engine also) to verify the safety of external messages.
+
+## Fault Proof implementation
+
+As described in the implementation-ideas, the fault-proof work of the superchain-backend reduces to accumulation of
+event-log hashes and aggregate checking of the initiating messages, at each L2 block height traversed in the proof.
+This fits in the backend interface as follows:
+- Cross-unsafe level safety is assumed for the previous block-height,
+  as safety checks are interleaved with regular processing one step at a time, it is all synchronous.
+  The "horizon" of new L2 blocks to cross-verify is essentially extended by 1 block at a time, to keep proving simple.
+- The executing and initiating message checks need to be in order of event-log,
+  to prevent any intra-block issues like cycles and time/ordering related cross-L2 tricks
+  (not enforcing intra-block ordering would enable arbitrary flashloan like behavior).
+
+## Integration into op-node
+
+After the completion of the
+[event-driven driver](https://github.com/ethereum-optimism/design-docs/blob/main/op-node-derivers.md)
+it should be possible to plug in the superchain-backed as an asynchronous verifier in the Driver module of the op-node.
+
+The backend events would roughly consist of:
+- Consume events:
+    - New unsafe block
+    - Reorg (if we want to support flip-flop reorgs, and not just wait for new blocks)
+- Produce events:
+    - Cross-unsafe reorg
+    - Unsafe to cross-unsafe promotion
+
+
+# Risks & Uncertainties
+
+<!-- *An overview of what could go wrong. Also any open questions that need more work to resolve.* -->
+
+## Throughput
+
+The embedded version in the op-node is too high latency:
+interop messaging may be faster or higher-throughput than an inefficient standard-RPC based
+implementation without prefetching or aggregating can keep up with.
+
+We may introduce non-standard RPC methods,
+to avoid overhead like fully encoding the receipts and hydrating their metadata.
+The event-log-filter query is already as narrowly scoped as possible (single block / contract),
+but is still effectively backed by receipts-retrieval in op-geth.
+
+If this proves to be too slow, then there still is the standalone service that benefits from prefetching
+and aggregate data handling (no receipts, just efficient message-hash lookups).
+In the long-term, with the growth of interoperating chains, light-clients (for just the cross-L2 part)
+may support the under-resourced nodes that are unable to run this service
+in addition to the current rollup-node and execution-engine stack.
```

Comment:
Eventually we will need to move to a more optimized protocol rather than RPC over HTTP 1